scvi/model/base/_demixin.py

```diff
@@ -0,0 +1,453 @@
+            logger.debug("n cells chunk {}".format(chunk.shape[0]))
+            res.append(
+                self._inference_loop(
+                    adata=adata,
+                    indices=chunk,
+                    n_samples=n_samples_per_cell,
+                    batch_size=batch_size,
+                    marginal_n_samples_per_pass=marginal_n_samples_per_pass,
+                    n_mc_samples_px=n_mc_samples_px,
+                )["hs_weighted"].numpy()
+            )
+            logger.debug(res[-1].shape)
+        res = np.concatenate(res, 0)
+        idx = np.arange(len(res))
+        idx = np.random.choice(idx, size=n_samples_overall, replace=True)
+        res = res[idx]
+        return res
+
+    @torch.no_grad()
+    def _inference_loop(
+        self,
+        adata: AnnData,
+        indices: Sequence,
+        n_samples: int,
+        marginal_n_samples_per_pass=500,
+        n_mc_samples_px=5000,
+        batch_size=64,
+    ):
+        """
+        Obtain gene expression and densities.
+
+        Computes gene normalized expression samples, as well as
+        variational posterior densities and likelihoods for each samples and each
+        cell.
+
+        The functioning of the method is a follows.
+        For each cell of the dataset, we sample `n_samples` posterior samples.
+        We then compute the likelihood of each sample and FOR EACH CELL, using
+        `_evaluate_likelihood`.
+        After concatenating likelihoods and samples, we derive overall
+        importance weights and importance weighted expression levels.
+
+        Parameters
+        ----------
+        adata : AnnData
+            Considered anndataset
+        indices : Sequence
+            Indices of the subpopulation
+        n_samples : int
+            Number of posterior samples per cell
+        marginal_n_samples_per_pass : int, optional
+            Number of samples per pass to compute the marginal likelihood, by default 500
+        n_mc_samples_px : int, optional
+            Number of overall samples per cell used to compute the marginal likelihood, by default 5000
+        batch_size : int, optional
+            Number of cells per minibatch, by default 64
+
+        Returns
+        -------
+        dict
+            Containing expression levels, z samples as well as associated densities
+        """
+        scdl = self._make_data_loader(
+            adata=adata, indices=indices, batch_size=batch_size, shuffle=False
+        )
+
+        zs = []
+        qzs_m = []
+        qzs_v = []
+        hs = []
+        log_px_zs = []
+        for tensors in scdl:
+            inference_outputs, generative_outputs, = self.module.forward(
+                tensors,
+                inference_kwargs=dict(n_samples=n_samples, return_densities=True),
+                compute_loss=False,
+            )
+            z = inference_outputs["z"].reshape(-1, self.module.n_latent).cpu()
+            h = generative_outputs["px_scale"]
```

Comment:
should we make px_scale an argument to this function so people can pick different h?